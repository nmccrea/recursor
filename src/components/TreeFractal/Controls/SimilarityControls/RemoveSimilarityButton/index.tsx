import React from "react"
import { useDispatch } from "react-redux"
import { removeOne } from "../../../../../state/similarities/actions"
import { SimilarityId } from "../../../../../state/similarities/types"
import Button from "react-bootstrap/Button"

interface RemoveSimilarityButtonProps {
  similarityId: SimilarityId
}

const RemoveSimilarityButton = ({
  similarityId,
}: RemoveSimilarityButtonProps) => {
  const dispatch = useDispatch()
  return (
    <Button
      variant="danger"
      onClick={() => {
        dispatch(removeOne(similarityId))
      }}
    >
      Remove
    </Button>
  )
}

export default RemoveSimilarityButton
