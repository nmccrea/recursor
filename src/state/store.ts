/**
 * Configures and exports the redux store and other top-level Redux resources.
 * @packageDocumentation
 */

import { configureStore, combineReducers } from "@reduxjs/toolkit"
import similarities from "./similarities/reducer"

const rootReducer = combineReducers({ similarities })
const createStore = () => configureStore({ reducer: rootReducer })

export { createStore }
export type RootState = ReturnType<typeof rootReducer>
export type Selector<ReturnType> = (state: RootState) => ReturnType | undefined
