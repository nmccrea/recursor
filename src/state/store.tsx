/**
 * Configures and exports the redux store and other top-level Redux resources.
 * @packageDocumentation
 */

import { combineReducers, configureStore } from "@reduxjs/toolkit"

const rootReducer = combineReducers({})
export type RootState = ReturnType<typeof rootReducer>
export default configureStore({ reducer: rootReducer })
