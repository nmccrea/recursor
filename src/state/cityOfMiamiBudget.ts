/**
 * TODO: type
 * TODO: test
 * TODO: doc
 */

import { RootState } from "./store"
import {
  createAsyncThunk,
  createSlice,
  CaseReducer,
  PayloadAction,
} from "@reduxjs/toolkit"

// State
enum Status {
  Idle = "idle",
  Pending = "pending",
  Fulfilled = "fulfilled",
}

interface CityOfMiamiBudgetState {
  status: Status
  data?: object
}

const INITIAL_STATE: CityOfMiamiBudgetState = { status: Status.Idle }

// Case Reducers
const fetchDataPending: CaseReducer<CityOfMiamiBudgetState, PayloadAction> = (
  previousState
) => ({
  ...previousState,
  status: Status.Pending,
})

const fetchDataFulfilled: CaseReducer<
  CityOfMiamiBudgetState,
  PayloadAction<object>
> = (previousState, action) => ({
  ...previousState,
  status: Status.Fulfilled,
  data: action.payload,
})

// Async Thunks
const fetchData = createAsyncThunk("fetchData", async (url: string) => {
  const response = await fetch(url)
  return await response.json()
})

// Slice
const { reducer } = createSlice({
  name: "cityOfMiamiBudget",
  initialState: INITIAL_STATE,
  reducers: {},
  extraReducers: {
    [fetchData.pending.type]: fetchDataPending,
    [fetchData.fulfilled.type]: fetchDataFulfilled,
  },
})

// Selectors
const getStatus = (state: RootState) =>
  (state.cityOfMiamiBudget && state.cityOfMiamiBudget.status) || undefined

const getData = (state: RootState) =>
  (state.cityOfMiamiBudget && state.cityOfMiamiBudget.data) || undefined

export default reducer
export { fetchData, getStatus, getData, Status }
