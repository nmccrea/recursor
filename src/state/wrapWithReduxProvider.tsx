/**
 * Implements Gatsby's `wrapRootElement` API.
 * @packageDocumentation
 */

import React from "react"
import { Provider as ReduxProvider } from "react-redux"
import store from "./store"

interface Arg {
  element: React.ReactNode
}
export const wrapWithReduxProvider = ({ element }: Arg) => (
  <ReduxProvider store={store}>{element}</ReduxProvider>
)
