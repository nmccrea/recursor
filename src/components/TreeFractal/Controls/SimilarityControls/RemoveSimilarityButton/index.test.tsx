import React from "react"
import { Provider } from "react-redux"
import { render, fireEvent, getByText } from "@testing-library/react"
import { createStore, RootState } from "../../../../../state/store"
import RemoveSimilarityButton from "."

describe("`<RemoveSimilarityButton />`", () => {
  it("renders correctly", () => {
    const state: RootState = {
      similarities: {
        ids: ["test/subject"],
        entities: {
          "test/subject": {
            id: "test/subject",
            translation: 0.7345,
            scale: 0.1672,
            angle: 0.7532,
            depth: 6,
            color: "violet",
          },
        },
      },
    }

    const { container } = render(
      <Provider store={createStore(state)}>
        <RemoveSimilarityButton similarityId={"test/subject"} />
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })

  it("responds correctly to a click event", () => {
    const state: RootState = {
      similarities: {
        ids: ["test/subject", "test/control"],
        entities: {
          "test/subject": {
            id: "test/subject",
            translation: 0.7345,
            scale: 0.1672,
            angle: 0.7532,
            depth: 6,
            color: "violet",
          },
          "test/control": {
            id: "test/control",
            translation: 0.2468,
            scale: 0.2856,
            angle: 0.224,
            depth: 7,
            color: "yellow",
          },
        },
      },
    }
    const store = createStore(state)
    const { container } = render(
      <Provider store={store}>
        <RemoveSimilarityButton similarityId={"test/subject"} />
      </Provider>
    )

    fireEvent.click(getByText(container, "Remove"))

    expect(store.getState()).toEqual<RootState>({
      similarities: {
        ids: ["test/control"],
        entities: {
          "test/control": {
            id: "test/control",
            translation: 0.2468,
            scale: 0.2856,
            angle: 0.224,
            depth: 7,
            color: "yellow",
          },
        },
      },
    })
  })
})
