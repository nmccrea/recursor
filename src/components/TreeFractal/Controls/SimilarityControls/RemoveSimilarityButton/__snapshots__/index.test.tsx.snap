// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`<RemoveSimilarityButton />\` renders correctly 1`] = `
<div>
  <button
    class="sc-AxjAm sc-AxiKw dWhrYK"
  >
    Remove
  </button>
</div>
`;
